src/components/Microsoft.DotNet.UpgradeAssistant.MSBuild/MSBuildWorkspaceUpgradeContext.cs

```diff
@@ -140,20 +143,110 @@ public IEnumerable<IProject> Projects
+                }
+
+                File.CreateSymbolicLink(target, entry);
+            }
+        }
+
+        private static string? GetMacOSMSBuildExtensionsPath(WorkspaceOptions options)
+        {
+            const string DefaultDotnetSdkLocation = "/usr/local/share/dotnet/sdk/";
+
+            if (options.MSBuildPath == null || !options.MSBuildPath.StartsWith(DefaultDotnetSdkLocation, StringComparison.Ordinal))
+            {
+                return null;
+            }
+
+            string? msbuildExtensionsPath = null;
+
+            if (Directory.Exists(MacOSMonoFrameworkMSBuildExtensionsDir))
+            {
+                // Check to see if the specified MSBuildPath contains the Mono.framework build extensions.
+                var monoExtensionDirectories = Directory.GetDirectories(MacOSMonoFrameworkMSBuildExtensionsDir);
+                var createTempExtensionsDir = false;
+
+                foreach (var monoExtensionDir in monoExtensionDirectories)
+                {
+                    var dotnetExtensionDir = Path.Combine(options.MSBuildPath, Path.GetFileName(monoExtensionDir));
+                    if (!Directory.Exists(dotnetExtensionDir))
+                    {
+                        createTempExtensionsDir = true;
+                        break;
+                    }
+                }
+
+                // If the specified MSBuildPath does not contain the Mono.framework build extensions, create a temp
+                // directory that we'll use to symlink everything.
+                if (createTempExtensionsDir)
+                {
+                    var homeDir = Environment.GetFolderPath(Environment.SpecialFolder.UserProfile);
+                    var versionDir = Path.GetFileName(options.MSBuildPath.TrimEnd('/'));
+
+                    msbuildExtensionsPath = Path.Combine(homeDir, ".dotnet-upgrade-assistant", "dotnet-sdk", versionDir);
+
+                    try
+                    {
+                        if (!Directory.Exists(msbuildExtensionsPath))
+                        {
+                            Directory.CreateDirectory(msbuildExtensionsPath);
+                        }
+
+                        // First, create symbolic links to all of the dotnet MSBuild file system entries.
+                        CreateSymbolicLinks(msbuildExtensionsPath, options.MSBuildPath);
+
+                        // Then create the symbolic links to the Mono.framework/External/xbuild system entries.
+                        CreateSymbolicLinks(msbuildExtensionsPath, MacOSMonoFrameworkMSBuildExtensionsDir);
+                    }
+#pragma warning disable CA1031 // Do not catch general exception types
+                    catch (Exception)
+#pragma warning restore CA1031 // Do not catch general exception types
+                    {
+                        // Something failed so we may be better off just using the default dotnet MSBuildExtensionsPath.
+                        msbuildExtensionsPath = null;
```

Comment:
Should we handle access/IO exceptions better and surface them to the user? E.g. do the need sudo permissions, etc.?

What about telemetry?

Reply:
The user won't need sudo unless they've broken their home directory permissions or something.

I've removed the catch-all and am now allowing these exceptions to surface to the user (& telemetry).